src/Compilers/CSharp/Portable/Symbols/Metadata/PE/TupleTypeDecoder.cs

```diff
@@ -0,0 +1,281 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Roslyn.Utilities;
+using System;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using System.Reflection.Metadata;
+
+namespace Microsoft.CodeAnalysis.CSharp.Symbols.Metadata.PE
+{
+    /// <summary>
+    /// In C#, tuples can be represented using tuple syntax and be given
+    /// names. However, the underlying representation for tuples unifies
+    /// to a single underlying tuple type, System.ValueTuple. Since the
+    /// names aren't part of the underlying tuple type they have to be
+    /// recorded somewhere else.
+    /// 
+    /// Roslyn records tuple names in an attribute: the
+    /// TupleElementNamesAttribute. The attribute contains a single string
+    /// array which records the names of the tuple elements in a breadth-first
+    /// traversal. If the type contains nested parameters, they are also recorded
```

Comment:
This implementation traverses children to the left completely before children to the right - that is depth-first. It also serializes parent before the children - that is pre-order.
BF (level-wise) traversal would put  "x", "y" in   `((int a, (int x, int y) b) c, (int d, int e) f)`  as the last ones since they are the deepest. In the given implementation , however, they would be recorded before "d", "e".

---
In reply to: [69969712](https://github.com/dotnet/roslyn/pull/12344#discussion_r69969712) [](ancestors = 69969712)